merge.py

```diff
@@ -1,98 +1,196 @@
-  print("Rebased")
-
-  # Check out target branch to here
-  checkout_success = call(['git', 'checkout', '{}/{}'.format(TARGET_REMOTE, TARGET_BRANCH)], cwd='/usr/local/google/home/jasonkuster/tmp/')
-  if not checkout_success == 0:
-    post_error('Error checking out target branch: master. Please try again.', pr)
-    return False
-  print("Checked out Apache master.")
-
-  # Merge
-  merge_success = call(['git', 'merge', '--no-ff', '-m', 'This closes #{}'.format(pr), 'finish-pr-{}'.format(pr)], cwd='/usr/local/google/home/jasonkuster/tmp/')
-  if not merge_success == 0:
-    post_error('Merge was not successful against target branch: master. Please try again.', pr)
-    return False
-  print("Merged successfully.")
-
-  print("Running mvn clean verify.")
-  # mvn clean verify
-  mvn_success = call(['mvn', 'clean', 'verify'], cwd='/usr/local/google/home/jasonkuster/tmp/')
-  if not mvn_success == 0:
-    post_error('mvn clean verify against HEAD + PR#{} failed. Not merging.'.format(pr), pr)
-    return False
-
-  # git push (COMMENTED UNTIL MERGEBOT HAS PERMISSIONS)
-  #push_success = call(['git', 'push', 'apache', 'HEAD:master'], cwd='/usr/local/google/home/jasonkuster/tmp/')
-  #if not push_success == 0:
-  #  post_error('Git push failed. Please try again.', pr)
-  #  return False
-  return True
-
-def set_up():
-  if not call(['mkdir', '/usr/local/google/home/jasonkuster/tmp']) == 0:
-    return False
-  return True
-
-
-def clean_up():
-  if not call(['rm', '-rf', '/usr/local/google/home/jasonkuster/tmp']) == 0:
-    return False
-  return True
 
+def create_merger(config, work_queue):
+    """create_merger creates a merger of the type specified in the config.
+
+    Args:
+        config: A dictionary containing repository configuration.
+        work_queue: A Queue.Queue where new work items will be queued.
+    Returns:
+        Merger of type specified in configuration
+    Raises:
+        AttributeError: if passed an unsupported SCM type.
+    """
+    if config['scm_type'] == 'github':
+        return GitMerger(config, work_queue)
+    raise AttributeError('Unsupported SCM type: {}.'.format(config['scm_type']))
+
+
+class Merger(object):
+    """Merger is the base class for all mergers.
+    """
+
+    def __init__(self, config, work_queue):
+        self.config = config
+        if not isinstance(work_queue, Queue.Queue):
+            raise AttributeError('Expected work_queue to be of type '
+                                 'Queue.Queue.')
+        self.work_queue = work_queue
+
+    def merge(self):
+        """merge spins forever, merging things off of the work queue.
+        """
+        raise NotImplementedError('merge is required to be implemented.')
+
+
+class GitMerger(Merger):
+    """GitMerger merges Git pull requests.
+    """
+
+    def merge(self):
+        """merge spins forever, merging things off of the work queue.
+
+        Raises:
+            AttributeError: if work items are not of type github_helper.GithubPR
```

Comment:
Should the exception be mentioned in the base class? Is there an equivalent of "inherit the parent documentation"? Is there any way to mark this method as abstract so it must be implemented?